host/ocalls.c

```diff
@@ -1,9 +1,35 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "ocalls.h"
-#include <openenclave/internal/calls.h>
 #include <stdio.h>
+#include <stdlib.h>
+
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/trace.h>
+
+#include "ocalls.h"
+
+void HandleMalloc(uint64_t arg_in, uint64_t* arg_out)
+{
+    if (arg_out)
+        *arg_out = (uint64_t)malloc(arg_in);
+}
+
+void HandleRealloc(uint64_t arg_in, uint64_t* arg_out)
+{
+    oe_realloc_args_t* args = (oe_realloc_args_t*)arg_in;
+
+    if (args)
+    {
+        if (arg_out)
+            *arg_out = (uint64_t)realloc(args->ptr, args->size);
```

Comment:
Are you sure it's correct to skip the realloc if arg_out is NULL?  This looks suspicious

Reply:
Heh, I copy-pasted this from SGX-specific sources.

This is rather curious. It seems that the interface should force `arg_out` to never be `NULL`. Otherwise, If `realloc` returns a pointer to a place in memory other than that originally pointed to by `args->ptr`, the caller has no way of knowing that the memory block has moved.

Reply:
On second thought, there is no way to enforce the requirement that `arg_out` not be `NULL` within this function since `arg_out` is the only way to signal to the caller that an error occurred. So calling this with `arg_out = NULL` is a programming error. I'll look at how it is used.

Reply:
Right, you should delete line 24 and just dereference *arg_out so it will crash on a programming error.

Reply:
I like it :) .

Reply:
Done.